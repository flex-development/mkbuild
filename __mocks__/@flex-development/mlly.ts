/**
 * @file Mocks - @flex-development/mlly
 * @module mocks/flex-development/mlly
 * @see https://github.com/flex-development/mlly
 */

/**
 * [`@flex-development/mlly`][1] module type.
 *
 * [1]: https://github.com/flex-development/mlly
 */
type Actual = typeof import('@flex-development/mlly')

/**
 * `@flex-development/mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/mlly')

export const CONDITIONS = actual.CONDITIONS
export const RESOLVE_EXTENSIONS = actual.RESOLVE_EXTENSIONS
export const fillModules = vi.fn(actual.fillModules)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const getSource = vi.fn(actual.getSource)
export const readPackageJson = vi.fn(actual.readPackageJson)
export const resolveModule = vi.fn(actual.resolveModule)
export const resolveModules = vi.fn(actual.resolveModules)
export const toDataURL = vi.fn(actual.toDataURL)
export const toURL = vi.fn(actual.toURL)
