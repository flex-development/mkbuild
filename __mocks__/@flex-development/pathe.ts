/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
