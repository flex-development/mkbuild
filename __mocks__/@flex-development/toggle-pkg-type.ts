/**
 * @file Mocks - @flex-development/toggle-pkg-type
 * @module mocks/flex-development/toggle-pkg-type
 * @see https://github.com/flex-development/toggle-pkg-type
 */

export default vi.fn()
