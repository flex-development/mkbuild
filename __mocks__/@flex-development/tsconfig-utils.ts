/**
 * @file Mocks - @flex-development/tsconfig-utils
 * @module mocks/flex-development/tsconfig-utils
 * @see https://github.com/flex-development/tsconfig-utils
 */

/**
 * [`@flex-development/tsconfig-utils`][1] module type.
 *
 * [1]: https://github.com/flex-development/tsconfig-utils
 */
type Actual = typeof import('@flex-development/tsconfig-utils')

/**
 * `@flex-development/tsconfig-utils` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>(
  '@flex-development/tsconfig-utils'
)

export const ScriptTarget = actual.ScriptTarget
export const normalizeCompilerOptions = vi.fn(actual.normalizeCompilerOptions)
export const normalizeModuleResolution = vi.fn(actual.normalizeModuleResolution)
export const resolvePaths = vi.fn(actual.resolvePaths)
