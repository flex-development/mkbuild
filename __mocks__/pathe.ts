/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

import path, { type FormatInputPathObject } from 'node:path'

/**
 * Return the last portion of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} File name
 */
export const basename = vi.fn((p: string): string => path.basename(p))

/**
 * Returns the directory name of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Directory containing `p`
 */
export const dirname = vi.fn((p: string): string => path.dirname(p))

/**
 * Returns the extension of `p`, from the last `'.'` to end of string in the
 * last portion of the path.
 *
 * If there is no `'.'` in the last portion of the path or the first character
 * of it is `'.'`, an empty string will be returned.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Extension of `p`
 */
export const extname = vi.fn((p: string) => path.extname(p))

/**
 * Returns a path string from an object - the opposite of {@link parse}.
 *
 * @param {FormatInputPathObject} p - Path object to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const format = vi.fn((p: FormatInputPathObject) => path.format(p))

/**
 * Parses `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const parse = vi.fn((p: string) => path.parse(p))

/**
 * Returns the relative path from `from` to `to` based on the current working
 * directory.
 *
 * @param {string} from - Start path
 * @param {string} to - Destination path
 * @return {string} Relative path from `from` to `to`
 */
export const relative = vi.fn((from: string, to: string): string => {
  return path.relative(from, to)
})

/**
 * Starting from leftmost {from} parameter, resolves {to} to an absolute path.
 *
 * @param {string[]} paths - Path segments
 * @return {string} Absolute path composed from `paths`
 */
export const resolve = vi.fn((...paths: string[]): string => {
  return path.resolve(...paths)
})
