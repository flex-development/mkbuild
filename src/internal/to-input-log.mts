/**
 * @file Internal - toInputLog
 * @module mkbuild/internal/toInputLog
 */

import type { ColorFunction, Colors } from '@flex-development/colors'
import type { InputLogObject } from '@flex-development/log'
import type { LogType, Message } from '@flex-development/mkbuild'
import pathe from '@flex-development/pathe'
import { fallback } from '@flex-development/tutils'
import { ok } from 'devlop'

/**
 * Convert a build message to an input log object.
 *
 * @see {@linkcode InputLogObject}
 * @see {@linkcode Message}
 *
 * @internal
 *
 * @this {void}
 *
 * @param {Message} msg
 *  Build message
 * @param {Colors} colors
 *  Color function map
 * @return {InputLogObject}
 *  Input log object
 */
function toInputLog(this: void, msg: Message, colors: Colors): InputLogObject {
  ok(typeof msg.level === 'string', 'expected `msg.level`')

  /**
   * Format arguments.
   *
   * @const {unknown[]} args
   */
  const args: unknown[] = []

  /**
   * Module id.
   *
   * @const {string | undefined} id
   */
  const id: string | undefined = msg.id || msg.loc?.file

  /**
   * Indentation.
   *
   * @const {string} indent
   */
  const indent: string = ' '.repeat(2)

  /**
   * Tint map.
   *
   * @const {Record<LogType, ColorFunction>} tint
   */
  const tint: Record<LogType, ColorFunction> = {
    debug: colors.gray,
    error: colors.red,
    info: colors.cyan,
    trace: colors.gray,
    verbose: colors.gray,
    warn: colors.yellow
  }

  /**
   * Message text.
   *
   * @var {string} message
   */
  let message: string = `${colors.bold('%s')} ${msg.message}`

  if (id) {
    message += `\n${indent}%s`
    args.push(colors.gray(pathe.relative(pathe.cwd(), id)))
  }

  if (msg.loc) {
    message += '%s'
    args.push(colors.gray(`:${msg.loc.line}:${msg.loc.column}`))

    if (msg.frame) {
      /**
       * Space to add before {@linkcode msg.loc.line}, as well as before
       * {@linkcode msg.frame}.
       *
       * @const {string} space
       */
      const space: string = indent.repeat(2)

      message += colors.gray(`:\n${space}%j |${space}`)
      message += colors.dim(tint[msg.level]('%s'))

      args.push(msg.loc.line, msg.frame.trimStart())
    }
  }

  return {
    args: [tint[msg.level](`[${fallback(msg.plugin, 'rollup')}]`), ...args],
    level: msg.level,
    message,
    stack: msg.stack,
    type: msg.level
  }
}

export default toInputLog
