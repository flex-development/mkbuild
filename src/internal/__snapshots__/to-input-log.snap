// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:internal/toInputLog > should return input log object (0) 1`] = `
{
  "args": [
    "[33m[rollup][39m",
  ],
  "level": "warn",
  "message": "[1m%s[22m Generated an empty chunk: "interfaces/task".",
  "stack": undefined,
  "type": "warn",
}
`;

exports[`unit:internal/toInputLog > should return input log object (1) 1`] = `
{
  "args": [
    "[31m[mkbuild:dts][39m",
    "[90msrc/utils/runnable-task.mts[39m",
    "[90m:449:17[39m",
    449,
    "formatLog('error', e)",
  ],
  "level": "error",
  "message": "[1m%s[22m Argument of type '"error"' is not assignable to parameter of type 'LogLevel'.
  %s%s[90m:
    %j |    [39m[2m[31m%s[39m[22m",
  "stack": RollupError: Argument of type '"error"' is not assignable to parameter of type 'LogLevel'.
  at getRollupError (file://\${process.cwd()}/node_modules/rollup/dist/es/shared/parseAst.js:396:41)
  at error (file://\${process.cwd()}/node_modules/rollup/dist/es/shared/parseAst.js:392:42)
  at Object.diagnostics (\${process.cwd()}/src/plugins/dts.mts:135:22)
  at Object.renderStart (\${process.cwd()}/src/plugins/dts.mts:359:17)
  at async Promise.all (index 0)
  at async PluginDriver.hookParallel (file://\${process.cwd()}/node_modules/rollup/dist/es/shared/node-entry.js:20868:9)
  at async Bundle.generate (file://\${process.cwd()}/node_modules/rollup/dist/es/shared/node-entry.js:19200:13)
  at async file://\${process.cwd()}/node_modules/rollup/dist/es/shared/node-entry.js:21904:27
  at async catchUnfinishedHookActions (file://\${process.cwd()}/node_modules/rollup/dist/es/shared/node-entry.js:21286:16),
  "type": "error",
}
`;

exports[`unit:internal/toInputLog > should return input log object (2) 1`] = `
{
  "args": [
    "[36m[mkbuild:esbuild][39m",
    "[90msrc/utils/load-build-config.mts[39m",
    "[90m:96:18[39m",
    96,
    ": (await import(url.href) as { default?: Config | null }).default",
  ],
  "level": "info",
  "message": "[1m%s[22m This "import" expression will not be bundled because the argument is not a string literal
  %s%s[90m:
    %j |    [39m[2m[36m%s[39m[22m",
  "stack": undefined,
  "type": "info",
}
`;
